### ISSUE TITLE
Incorrect data fetch by Civetweb client from Civetweb server

### ISSUE THREAD

**Issue by daokoder** [bel2125/civetweb#161]
This happens when the server sends the response *without* the `Content-Length` header, and sends the data in a *separate* `mg_printf()`. One can reproduce this issue by using the `hello` example as a server, but changing that single `mg_printf()` to the following two:
~~~c
    mg_printf(conn, "HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\n\r\n" );
    mg_printf(conn, "%s", content);
~~~
And for the client, one may use:
~~~c
int main(void)
{
    char postbuf[4*1024];
    char ebuf[1000];

    struct mg_connection *conn = mg_download(
            "localhost", 8080, 0, ebuf, sizeof(ebuf),
            "GET / HTTP/1.1\r\nHost: localhost\r\n"
            "Content-type: text/plain\r\n"
            "Content-length: 0\r\n\r\n" );

    if( conn == NULL ) return 1;

    while(1){
        int postlen = mg_read( conn, postbuf, sizeof(postbuf)-1 );
        if( postlen <= 0 ) break;
        printf( "%s", postbuf );
    }   
    mg_close_connection( conn );
    fflush( stdout );
    return 0;
}
~~~
The test indicates that the client often doesn't read data from the server. Please note that Mongoose (commit 04fc209644b4, Fri Aug 16 11:18:04 2013) was working properly on this test.

**Comment by bel2125** [comment]
I think

    mg_printf(conn, "HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\n\r\n" );
    mg_printf(conn, "%s", content);

is not correct. You use `HTTP/1.1` which implies `Connection: Keep-Alive` as a default option, in contrast to `HTTP/1.0` (see https://en.wikipedia.org/wiki/HTTP_persistent_connection#Operation). You also do not set a `Content-Length` header, which is mandatory for "keep-alive" connections.
This way, the client has no chance to determine if the content is complete.

You must either set `Connection: close`, use `HTTP/1.0` or set `Content-Length: <content_length>`, otherwise it is an HTTP protocol violation.
You can use the fiddler2 web debugging tool and turn on the protocol checks. It will report this issue, and additionally it will report that a date should be sent from the server.

It may be true that an old (or even a new) version of mongoose did work better on this test - the test itself is not correct! 
Did you enable keep-alive in the server?

**Comment by bel2125** [comment]
I would think, the difference between one line `mg_printf( <header + content> )` and the two lines above is because the first one will generate one Ethernet frame the second one two. 
The way these frames are generated can be influenced by socket options like TCP_NODELAY and TCP_CORK (http://baus.net/on-tcp_cork/, http://www.techrepublic.com/article/tcp-ip-options-for-high-performance-data-transmission/).

Still, this is an explanation why you might see a difference. The problem still remains the HTTP protocol violation. If the protocol is used correctly, you will not see any difference on the functional point of view - these TCP options will only influence performance, not functionality.

**Comment by Night-walker** [comment]
In the scenario we are testing, keep-alive is not enabled and the connection is closed by the server after responding, as otherwise the client side would hang on `recv()`. In this case `Content-length` in the response is optional according to RFC 7230, but the client fails to properly determine it. `Connection: close` put in the request yields no effect, of course.

**Comment by daokoder** [comment]
>  You use HTTP/1.1 which implies Connection: Keep-Alive as a default option, in contrast to HTTP/1.0 

I just tried with `HTTP/1.0`, the problem remains the same. FYI, the tested platform is Mac OSX 10.10.1 (I did try a couple of times on Lubuntu 14.04, seems ok).

**Comment by bel2125** [comment]
I don't have a Mac OS platform, so I can not test it directly. I can only test in Windows and Linux.
What client are you using? Are you using a browser, are you using civetweb as in the example above, or something else?
If you use civetweb as a client: I recently made some changes in `mg_read`.

**Comment by daokoder** [comment]
> What client are you using? Are you using a browser, are you using civetweb as in the example above, or something else?

The client uses civetweb as in the example. The problem may be in the client part, I tried with a web browser and wget, neither have such problem.

> If you use civetweb as a client: I recently made some changes in mg_read.

I tried both the latest commit (as of yesterday) and the one in release V1.7, both have the problem. I see you have made some new commits, I will try that as well.

**Comment by Night-walker** [comment]
> I don't have a Mac OS platform, so I can not test it directly. I can only test in Windows and Linux.

I experienced this issue both on Windows 7 and Ubuntu 14.04.

**Comment by bel2125** [comment]
The changes from a few hours ago should make the client more robust against this protocol violation.
Still, you should use chunked transfer encoding - that is the standard way if you do not know in advance how much data has to be sent.
I will add an example client soon.

**Comment by Night-walker** [comment]
It is not a violation of HTTP protocol, it is an acceptable case in both RFC 2616 and RFC 7230.

**Comment by bel2125** [comment]
Fiddler states:

Fiddler has detected a protocol violation in session # xxx. No Connection: close, no Content-Length. No way to tell if the response is complete.

However, searching the web for this error message, it seems such replies are sent by some servers sometimes. So the client should be able to work with such messages - no matter if they are protocol violations (like fiddler states) or not.

**Comment by bel2125** [comment]
The CivetWeb mg_download API is rather limited, and only remains for backward compatibility. There is a new client API, which gives full control, and also allows keep-alive, POST requests with chunked transfer encoding:

```C
/* First connect to the server */
conn = mg_connect_client("localhost", port, use_ssl, ebuf, sizeof(ebuf));

/* Send the header and terminate the header with an empty line */
mg_printf(conn, "POST /somewhere.cgi HTTP/1.1\r\n");
mg_printf(conn, "Host: www.example.com\r\n");
mg_printf(conn, "Transfer-encoding: chunked\r\n");
mg_printf(conn, "\r\n");

/* While there are chunks of data that need to be sent from the
   client to the server, send chunk sizes and chunks. */
while ((chunk = GetNextChunk()) != NULL) {
    mg_printf(conn, "%x\r\n%s\r\n", strlen(chunk), chunk);
}

/* Terminate the body data send from client to server */
mg_printf(conn, "0\r\n\r\n");

/* Now get the response from the server */
ok = mg_get_response(conn, ebuf, sizeof(ebuf), 1000) > 0;

/* Now call mg_read ... same as before */
...
```

**Comment by shawnye** [comment]
That's great, but `mg_get_response` can be used by Lua?

**Comment by bel2125** [comment]
Currently the client API is not available in Lua, neither the old one (`mg_download`) nor the new one (`mg_connect_client` + `mg_get_response`). But of course I can add it. Also the websocket client API (`mg_connect_websocket_client`) is not yet available in Lua.

However, the `connect` function is available in Lua already since a while. It works on TCP level and must be used like this (I did not test this particular example):
```Lua
connection = connect(host, port, is_ssl) -- I think it does only work with is_ssl = 0
connection:send("POST /something.cgi HTTP/1.1\r\n")
connection:send( ... everything from above ...  )
response = connection:recv()
-- here you need to parse the full HTTP response
connection:close()
```

`connect` is the only function outside the `mg` namespace - it was there before I took over CivetWeb and I kept it for compatibility reasons. C did not have a client API like Lua `connect` before, so the Lua API could do more than the C API. The Lua `connect` function is still even lower level.
